scenedetect/detectors/hash_detector.py

```diff
@@ -0,0 +1,168 @@
+        frame to frame.
+
+        Arguments:
+            frame_num (int): Frame number of frame that is being passed.
+
+            frame_img (Optional[int]): Decoded frame image (numpy.ndarray) to perform scene
+                detection on. Can be None *only* if the self.is_processing_required() method
+                (inhereted from the base SceneDetector class) returns True.
+
+        Returns:
+            List[int]: List of frames where scene cuts have been detected. There may be 0
+            or more frames in the list, and not necessarily the same as frame_num.
+        """
+
+        cut_list = []
+        metric_keys = self._metric_keys
+        _unused = ''
+
+        # Initialize last scene cut point at the beginning of the frames of interest.
+        if self.last_scene_cut is None:
+            self.last_scene_cut = frame_num
+        
+        # We can only start detecting once we have a frame to compare with.
+        if self.last_frame is not None:
+            # We obtain the change in hash value between subsequent frames as 
+            # well as the actual hash value. This is refered to in a statsfile
+            # as their respective metric keys.
+            if (self.stats_manager is not None and
+                    self.stats_manager.metrics_exist(frame_num, metric_keys)):
+                hash_dist = self.stats_manager.get_metrics(frame_num, metric_keys)
+            else:
+                # Perceptual hashing algorithm based on phash, updated to use OpenCV instead of PIL + scipy
+                # https://github.com/JohannesBuchner/imagehash
+                
+                # Convert to grayscale
+                curr_gray = cv2.cvtColor(frame_img, cv2.COLOR_BGR2GRAY)
+                # Resize image to square to help with DCT
+                imsize = self.hash_size * self.highfreq_factor
+                curr_resized = cv2.resize(curr_gray, (imsize, imsize), interpolation=cv2.INTER_AREA)
+                # Calculate discrete cosine tranformation of the image
+                curr_resized = numpy.float32(curr_resized) / numpy.max(numpy.max(curr_resized))
+                curr_dct = cv2.dct(curr_resized)
+                # Only keep the low frequency information
+                curr_dct_low_freq = curr_dct[:self.hash_size, :self.hash_size]
+                # Calculate the median of the low frequency information
+                curr_med = numpy.median(curr_dct_low_freq)
+                # Transform the low frequency information into a binary image based on > or < median
+                curr_hash = curr_dct_low_freq > curr_med
+                
+                last_hash = self.last_hash
+                
+                if last_hash.size == 0:
+                    # Calculate hash as above
+                    last_gray = cv2.cvtColor(self.last_frame, cv2.COLOR_BGR2GRAY)
+                    last_resized = cv2.resize(last_gray, (imsize, imsize), interpolation=cv2.INTER_AREA)
```

Comment:
Should refactor these calculations into a static function called "calculate_frame_hash" or something, just to reduce code duplication w/ above.

Reply:
Done with 0f54efc.